File: langs/swift/Sources/VMx/Messages/FormRevertedMessage.swift
//
// FormRevertedMessage — emitted when a FormVM reverts its model to snapshot.
//
// See spec/20-form-vm.md §8 — Hub messages.
//
import Foundation

/// Hub envelope broadcast by ``FormVM`` when `denyCommand` is executed and
/// the live model is reverted to the current snapshot.
///
/// Mirrors `FormRevertedMessage` in the C# / Python / TS flavors.
public struct FormRevertedMessage: Message {
    public let senderObject: AnyObject
    public let senderName: String

    public init(senderObject: AnyObject, senderName: String) {
        self.senderObject = senderObject
        self.senderName = senderName
    }

    /// Spec alias of `senderObject`. Provided to match the C# / Python
    /// `Sender` / `sender` field name.
    public var sender: AnyObject { senderObject }
}

File: langs/csharp/src/VMx/Forms/FormRevertedMessage.cs
using VMx.Messages;

namespace VMx.Forms;

/// <summary>
/// Published on the message hub when a <see cref="FormVM{TM}"/> reverts its
/// <c>Model</c> to <c>Snapshot</c> via <c>DenyCommand</c>.
///
/// See spec/20-form-vm.md §8 — Hub messages.
/// </summary>
/// <param name="Sender">The <see cref="FormVM{TM}"/> that was reverted.</param>
/// <param name="SenderName">Human-readable type name of the sender.</param>
public sealed record FormRevertedMessage(object Sender, string SenderName) : IMessage
{
    /// <inheritdoc/>
    public object SenderObject => Sender;
}
